### PR TITLE
Improve multicursor support

### LINKED ISSUE #1946: Completion error with multiple cursor
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Environment**:

- **VSCode Version**: 1.14.2 - cb82febafda0c8c199b9201ad274e25d9a76874e
- **VsCodeVim Version**: v0.9.0 Hundredth Hazelnut (Jun 24, 2017)
- **OS**: GNU/Linux

**What happened**:

Bug/error when completing go code

`for<tab>`

gives me : 

```
for index := 0; index < count; index++ {
	
}
```

with multiple cursors around "index"

pressing "backspace", to change value of "index" gives me : 

```
for ex < count; index++ {
	
}
```
**What did you expect to happen**: 

changing the index value to something I want. If I put VSCodeVim off, everything is ok.

**How to reproduce it**:

Ensure Plugings : 

1. go - 0.6.63 - 26th July, 2017
2. VSCodeVim
3. Open a .go file
4. Expand a "basic" for snippet

I'm not sure if it's related to VSCodeVim or go snippets. Again, if I turn VSCodeVim off, everything is ok.

### LINKED ISSUE #3005: breaks use of snippet placeholders
When entering the variable parts of a snippet, the placeholder for the current part should be automatically replaced by whatever is typed. This works when the Vim extension is disabled. When it is enabled, the entered text just goes after the placeholder and the placeholder remains. The end result is that you really can't use snippet placeholders when the Vim extension is enabled.

This only seems to happen when the placeholder variable is referred to again on another line. For example,

```json
  "test": {
    "prefix": "test",
    "body": ["line 1 ${1:my-placeholder}", "placeholder = ${1}"]
  },
```

### LINKED ISSUE #3976: Multiple cursor backspace affect only one line when using snippets
When I try to delete chars with backspace in same placeholders backspace affect only first one and also delete next chars (" =" in this example).
Alt Text

I disabled all extensions and after this, all worked right and when only vim is enabled this error appear.
This is the snippet that I use:
```json
"For loop": {
	"prefix": "for",
	"body": [
		"for(${1:int} ${2:i} = ${3:0}; ${2:i} < ${4:count}; ${2:i}++) {",
		"\t${0:/*code*/}",
		"}" 
	],
}
```

- Extension (VsCodeVim) version: v1.9.0 (2019-07-29)
- VSCode Version: 1.37.0
- OS Version: Windows 10

### LINKED ISSUE #4223: Expanding the snippet with multicursor is broken
**Describe the bug**
I am using the VSCodeVim/Vim plugin with the C/C++ plugin by Microsoft. When I enter `for<tab>`, the plugin expands the for-loop snippet. The behavior of modifying the iterating variable for this snippet seems to be broken. When I type `j` to replace `i`, the iterating variable becomes `ij` instead of `j`. I also could not delete the text properly.

<a href="https://gyazo.com/15c3958ac18f74e195c987849d7ceada">Image from Gyazo</a>

**To Reproduce**
1. Install the C/C++ plugin by Microsoft
2. Open a new C++ file
3. Type `for<tab>`
4. Type `<tab>` to select the iteration variable `i`
5. Try to modify the variable

**Expected behavior**
As I mentioned in the **Describe the bug** section.
Also, I verified that this behavior did not happen when I disable the Vim plugin.

**Screenshots**
As I mentioned in the **Describe the bug** section.

**Environment (please complete the following information):**
- VSCode (v.1.39.2)
- Vim plugin (v1.11.3)
- C/C++ plugin by Microsoft (v0.26.0)
- OS: macOS Mojave (v10.14.6)

**Additional context**
None.

### LINKED ISSUE #4281: When write c/c++ code use the 'snippets'，tab switch the key words，backspace is not doing well with the Multi-cursor.But when i disable Vim emulation, it work fine.
**Describe the bug**

**To Reproduce**

1.  Go to '...'
2.  Click on '....'
3.  Scroll down to '....'
4.  See error

**Expected behavior**

**Screenshots**
If remapping-related, please attach log output: https://github.com/VSCodeVim/Vim#debugging-remappings.

**Environment (please complete the following information):**

- Extension (VsCodeVim) version:
- VSCode version:
- OS:

**Additional context**

### LINKED ISSUE #4515: Autocomplete disappears after typing
Issue Type: <b>Bug</b>

**Describe the bug**
Autocomplete dissappears after typing in LaTex begin environments. Works after disabling vim.

**To Reproduce**
1.) Create basic latex file e.g.:
````
\documentclass{article}
\begin{document}

\end{document}
````

2.) Inside document type \begin and press tab to autocomplete begin environment

3.) Type any characters and autocomplete will disappear.

**Expected behavior**
Autocomplete should stay open and filter suggestions as you type.

**Screenshots**
Behavior with vim enabled:
vim

Behavior without vim enabled:
no_vim

Extension version: 1.12.4
VS Code version: Code 1.41.1 (26076a4de974ead31f97692a0d32f90d735645c0, 2019-12-18T14:58:56.166Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-2500K CPU @ 3.30GHz (4 x 3300)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|23.95GB (10.48GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

### LINKED ISSUE #4522: Consider using VSCode's type command to improve multicursor support
**Is your feature request related to a problem? Please describe.**
There are many issues with the extension's poor multicursor support, but one of the most annoying ones is that it breaks snippets and auto-completion. On single cursor mode the extension passes over the text to the `default: type` command through `TextEditor.insert()`, this way vscode has control over the text input and can do stuff like auto-completion, snippet mode, closing brackets, etc. 

Multiple cursors, however, are handled differently. `vscode.TextEditorEdit.insert()` is used instead, which doesn't allow vscode to do its regular stuff that it'd do on text insertion by default. 

**Describe the solution you'd like**
Use the `default: type` command to process text insertion on multicursor mode like it is used on single cursor mode.

**Describe alternatives you've considered**
Here is a rudimentary prototype of my suggestion that solves issues like #4281, #4515, #4223.

```ts
if (this.vimState.isMultiCursor
      && otherTransformations[0]?.type === 'insertTextVSCode'
      && otherTransformations.every(t => t?.type === 'insertTextVSCode')) {

        const { text } = otherTransformations[0];
        vscode.commands.executeCommand('default:type', { text });

} else {
// The regular way we handle `otherTransformations`
}
```
VSCode's `type` command handles replicating what happens at the first cursor on the other cursors, and unless I'm wrong we can't do things like place `x` at cursor[0] and `y` at cursor[1] so there is no need to create an edit for each individual cursor as long as we know they're all the same.

I believe a better solution would be to introduce a new transformation (`MulticursorInsert`) that can know the cursor position like `InsertTextTransformation`, but is processed similarly to an `InsertTextVSCodeTransformation`. 

**Additional context**
Before getting any further into it I'd like to know if this would be a welcomed change. I could submit a WIP PR if that's the case.

### PR DESCRIPTION
**Pull request by cvaldev**
**What this PR does / why we need it**:
It fixes multicursor snippets, and adds support for stuff like auto-bracketing and auto-completion on multicursor mode.

5fd59f7: (Backspace Support)
Makes `range` represent the `selection` at `position`. This way we delete the correct `range` rather than the primary selection as it was before these changes.

59496be: (Multicursor Character Insert)
Makes multicursor text insertion behave like single cursor text insertion by ultimately using the `default: type` command rather than `vscode.TextEditor.insert()` which doesn't have support for snippet mode, auto completion, bracketing, etc.

To accomplish this, I've added an optional property (`isMultiCursor`) to `InsertTextVSCodeTransformation` which represents `vimState.isMultiCursor`at the time the transformation was created. This property is used to filter the multicursor transformations into  `multicursorTextTransformations`.

Because the `default: type` command handles inserting the text into the correct cursor positions there's no need to repeat the transformation multiple times and we can just execute the command once, as long as we know all transformations are trying to do the same thing. 

**Which issue(s) this PR fixes**
Fixes #4281
Fixes #4223
Fixes #3005 
Fixes #4522
Fixes #4515 
Fixes #1946
fix #3976

**Special notes for your reviewer**:
Is there a way to programmatically test this stuff?... I've pretty much just done a bunch of manual testing, and it all seems to work fine. Let me know what you think, I'd be happy to keep working on it.

### CODE REVIEW

**Comment by J-Fields** [comment]
> Is there a way to programmatically test this stuff?

Not easily. I've filed https://github.com/VSCodeVim/Vim/issues/4582, which I'll try to get to soon

File: src/transformations/transformations.ts [hunk 1]
```diff
@@ -330,3 +338,16 @@ export const areAnyTransformationsOverlapping = (transformations: TextTransforma
 
   return false;
 };
+
+export const areAllSameTransformation = (transformations: Transformation[]): Boolean => {
+  const firstTransformation = transformations[0];
+
+  return transformations.every(t => {
+    for (const [key, value] of Object.entries(firstTransformation)) {
```
Comment (J-Fields) [inline src/transformations/transformations.ts]: I think this for loop can also be written with `every`

**Comment by J-Fields** [comment]
I'm guessing this would also fix #3976?

File: src/transformations/transformations.ts [hunk 2]
```diff
@@ -291,6 +296,9 @@ export const isTextTransformation = (x: Transformation): x is TextTransformation
     x.type === 'moveCursor'
   );
 };
+export const isMultiCursorTextTransformation = (x: Transformation): Boolean => {
+  return (x.type === 'insertTextVSCode' && x.isMultiCursor) ?? false;
+};
 
 const getRangeFromTextTransformation = (transformation: TextTransformations): Range | undefined => {
```
Comment (J-Fields) [inline src/transformations/transformations.ts]: Is there a reason these aren't defined with `function`?

File: src/transformations/transformations.ts (hunk 1 above)
Comment (cvaldev) [reply src/transformations/transformations.ts]: Thanks, I'll get this fixed

File: src/transformations/transformations.ts (hunk 2 above)
Comment (cvaldev) [reply src/transformations/transformations.ts]: I was just trying to keep style consistency of the document and didn't think it'd make a difference. Should I change them?

Comment (J-Fields) [reply src/transformations/transformations.ts]: Up to you, just making sure I wasn't missing something

**Comment by J-Fields** [comment]
Sorry for the merge conflicts, by the way. Once those are resolved and I get the chance to play around with this a bit, we can merge it in. I'm focusing this release on getting multi-cursor to "just work" as much as possible, and this'll be a huge help. I'm going to write a whole bunch of tests once #4583 is done, so no need right now.

**Pull request merged**